File: sessions/2018/easy/104_maximum_depth_of_binary_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0

        q = [(root,1)]
        result = 1
        while len(q) > 0:
            node, l = q.pop()
            result = max(result, l)
            if node and node.left:
                q.append((node.left, l+1))
            if node and node.right:
                q.append((node.right, l+1))
        return result

File: pia/2020/easy/104_maximum_depth_of_binary_tree .js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function (root) {
  return explore(root, 0);
};

function explore(node, currDepth) {
  if (!node) {
    return currDepth;
  }
  let leftDepth = explore(node.left, currDepth + 1);
  let rightDepth = explore(node.right, currDepth + 1);
  return Math.max(leftDepth, rightDepth);
}
